### PR TITLE
server:  allow filtering llama server response fields

### PR DESCRIPTION
**Pull request by nvrxq**
This adds a new feature. Allow Filtering LLama Server Response Fields from #10819.
Usage:
Curl: 
```bash
curl --request POST \
    --url http://localhost:8080/completion \
    --header "Content-Type: application/json" \
    --data '{"prompt": "Building a website can be done in 10 simple steps:","n_predict": 128, "response_fields": ["content", "generation_settings/n_predict"]}'
```
Response:
```bash
{"content":" 1. Choose a domain name...","generation_settings/n_predict":128}
```

```bash
Retrievals from multiple nested lists are also supported, example:
Given the original data: {"key1" : {"key2": 3}}
With response_fields = ["key1/key2"]
Then, the final response will be: {"key1/key2": 3}.
```

If [response](https://github.com/ggerganov/llama.cpp/blob/5478bbcd173e7027af7689493c7421719f5c43df/examples/server/server.cpp#L512) does not contain a path, we will return empty json for path.

### CODE REVIEW

**Review by ngxson — CHANGES_REQUESTED** [review]
Please also include a guide in the `/completion` endpoint section of server's documentation.

Adding test case is optional, but is highly recommended. See an example in `examples/server/tests/unit/test_completion.py`

File: examples/server/tests/unit/test_completion.py [hunk 1]
```diff
@@ -257,6 +257,40 @@ def check_slots_status():
         # assert match_regex(re_content, res.body["content"])
 
 
+@pytest.mark.parametrize(
+    "prompt,n_predict,requested_fields",
+    [
+        ("I believe the meaning of life is", 8, []),
+        ("I believe the meaning of life is", 32, ["content", "generation_settings/n_predict", "prompt"]),
```
Comment (ngxson) [inline examples/server/tests/unit/test_completion.py:264]: What happen if I want the whole `"generation_settings"` object, not just some keys inside?

Comment (nvrxq) [reply examples/server/tests/unit/test_completion.py:264]: Return fully generation_settings. 
Example:
```bash
curl --request POST \
    --url http://localhost:8080/completion \
    --header "Content-Type: application/json" \
    --data '{"prompt": "Hi","n_predict": 128, "requested_fields": [ "generation_settings"]}'
 ```
 
 Response:
```bash
{"generation_settings":{"n_predict":128,"seed":4294967295,"temperature":0.800000011920929,"dynatemp_range":0.0,"dynatemp_exponent":1.0,"top_k":40,"top_p":0.949999988079071,"min_p":0.05000000074505806,"xtc_probability":0.0,"xtc_threshold":0.10000000149011612,"typical_p":1.0,"repeat_last_n":64,"repeat_penalty":1.0,"presence_penalty":0.0,"frequency_penalty":0.0,"dry_multiplier":0.0,"dry_base":1.75,"dry_allowed_length":2,"dry_penalty_last_n":4096,"dry_sequence_breakers":["\n",":","\"","*"],"mirostat":0,"mirostat_tau":5.0,"mirostat_eta":0.10000000149011612,"stop":[],"max_tokens":128,"n_keep":0,"n_discard":0,"ignore_eos":false,"stream":false,"logit_bias":[],"n_probs":0,"min_keep":0,"grammar":"","samplers":["penalties","dry","top_k","typ_p","top_p","min_p","xtc","temperature"],"speculative.n_max":16,"speculative.n_min":5,"speculative.p_min":0.8999999761581421,"timings_per_token":false,"post_sampling_probs":false}}
```

File: examples/server/README.md [hunk 2]
```diff
@@ -450,6 +450,8 @@ These words will not be included in the completion, so make sure to add them to
 
 `post_sampling_probs`: Returns the probabilities of top `n_probs` tokens after applying sampling chain.
 
+`requested_fields`: A list of required response fields, for example : `"requested_fields": ["content", "generation_settings/n_predict"]`  If there is no field, return an empty json for that field.
```
Comment (ngxson) [inline examples/server/README.md:453]: > If there is no field, return an empty json for that field.

Could you explain what does this mean? Please give an example if possible.

Comment (nvrxq) [reply examples/server/README.md:453]: Yep, sure. 
If a user requests a non-existent key, for example : “requested_fields”: [“content”, “generation_settings/SmthNonExistent”]`, then only “content” will be returned.  I was thinking about writing the absence of like in the response: 
{ “content” : “text”, “generation_settings/SmthNonExistent”:  “incorrect path"}, but haven't come to a consensus.

Comment (ngxson) [reply examples/server/README.md]: The current behavior is ok, but just need to be more clear when writing docs.

`return an empty json` means `{}`, it's not the current behavior. I rephrased this to be more clear.

**Review by ngxson — APPROVED** [review]

File: examples/server/README.md (hunk 2 above)
Comment (nvrxq) [reply examples/server/README.md]: Окей, понял. Спасибо!

**Comment by isaac-mcfadyen** [comment]
When filtering nested response fields, is this new option un-nesting them?

In the example provided:
```bash
curl --request POST \
    --url http://localhost:8080/completion \
    --header "Content-Type: application/json" \
    --data '{"prompt": "Building a website can be done in 10 simple steps:","n_predict": 128, "response_fields": ["content", "generation_settings/n_predict"]}'
```
```json
{"content":" 1. Choose a domain name...","generation_settings/n_predict":128}
```

It seems like it's changing the previous format of `{"generation_settings": {"n_predict": 128}}` (nested) to `{"generation_settings/n_predict": 128}` (un-nested, with the key having the slash instead).

Is this intended or am I misunderstanding somewhere?

**Comment by nvrxq** [comment]
I've thought about it, but again haven't come to any sort of consensus.
If the user in the conditional code will call response_fields, then in the script where we return {“generation_settings”: {“n_predict”: 128}} he would first have to do a split on response_fields (“/”), and then fetch the values.

**Comment by ngxson** [comment]
I don't have any preference either.

In a logical sense, if we only need pick some fields inside `generation_settings`, then you need to modify program's code anyway. So it's trivial to change your code to read `generation_settings/n_predict` instead of `generation_settings["n_predict"]`

Also I'm not sure if many users gonna use this functionality, so probably not urgent to change anything right now. Let's wait a bit more and see if it's worth changing.

**Comment by isaac-mcfadyen** [comment]
> Let's wait a bit more and see if it's worth changing

Alright, sounds good. I might create a PR just to add a quick docs update in case anyone runs into this when using this field if that works.

**Pull request merged**